dvc/repo/add.py

```diff
@@ -98,8 +98,11 @@ def add(  # noqa: C901
                 **kwargs,
             )
 
+            # remove existing stages that are to-be replaced with these
+            # new stages for the graph checks.
+            old_stages = set(repo.stages) - set(stages)
             try:
-                repo.check_modified_graph(stages)
+                repo.check_modified_graph(stages, list(old_stages))
```

Comment:
This would be better reflected by `Index` here, as we are creating a new index. Index is a collection of stages, so if some stages underneath are changed, is it the same index? So, immutability is at the heart of the index and would have been better reflected here.
```suggestion
            new_index = repo.index.update(stages)
            try:
                new_index.check_graph()
```